src/server.rs

```diff
@@ -69,6 +71,83 @@ fn replace_string_in_range(
     contents
 }
 
+fn function_defines(
+    name: String,
+    params: &Vec<FunctionParameter>,
+) -> bool {
+    for param in params {
+        if param.key.name == name {
+            return true;
+        }
+    }
+
+    false
+}
+
+fn is_scope(name: String, n: Rc<walk::Node<'_>>) -> bool {
+    let mut dvisitor = DefinitionFinderVisitor::new(name);
+    walk::walk(&mut dvisitor, n.clone());
+    let state = dvisitor.state.borrow();
+
+    state.node.is_some()
+}
+
+fn find_references(
+    uri: lsp::Url,
+    result: NodeFinderResult,
+) -> Vec<lsp::Location> {
+    let mut locations: Vec<lsp::Location> = vec![];
+
+    if let Some(node) = result.node {
+        let name = match node.as_ref() {
+            walk::Node::Identifier(ident) => Some(ident.name.clone()),
```

Comment:
I think you can avoid a clone here too. There's two ways and I'm not the most familiar with rust lifetimes to know if the first will work, but I think the second will.

First, I think that you don't need to use a clone here. I think you can return a reference so you'll have `Some(&ident.name)`. Since this reference never leaves this function, the lifetime should be ok and Rust shouldn't complain.

The second is I think this function exits immediately if the matcher that returns `None` is reached. I don't see any branch where it does anything meaningful on null. If this is the case, you can do this instead:

```
let name = match node.as_ref() {
    walk::Node::Identifier(&ident) => &ident.name,
    walk::Node::IdentifierExpr(&ident) => &ident.name,
    _ => return Vec::new(),
};

You can then access name directly as a `&str` rather than needing to start the next section with an `if let`.